python/dgl/backend/pytorch/sparse_emb.py

```diff
@@ -0,0 +1,418 @@
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def world_size(self):
+        return self._world_size
+
+    @property
+    def num_embeddings(self):
+        return self._num_embeddings
+
+    @property
+    def queues(self):
+        return self._queues
+
+    def set_opt_state(self, state):
+        self._state = state
+
+    @property
+    def opt_state(self):
+        return self._state
+
+    @property
+    def trace(self):
+        return self._trace
+
+    def reset_trace(self):
+        self._trace = []
+
+    @property
+    def emb_tensor(self):
+        return self._tensor
+
+class SparseGradOptimizer(abc.ABC):
+    r''' The abstract sparse optimizer.
+
+    Parameters
+    ----------
+    params : list of GraphSparseEmbedding
+        The list of GraphSparseEmbeddings.
+    lr : float
+        The learning rate.
+    '''
+    def __init__(self, params, lr):
+        self._params = params
+        self._lr = lr
+
+    def step(self):
+        ''' The step function.
+
+        The step function is invoked at the end of every batch to update embeddings
+        '''
+        with th.no_grad():
+            update_embs = {emb.name: ([],[]) for emb in self._params}
+            for emb in self._params:
+                num_embeddings = emb.num_embeddings
+                emb_name = emb.name
+                range_size = (num_embeddings + self._world_size - 1) // self._world_size \
+                    if self._world_size > 0 else 0
+                for idx, data in emb._trace:
+                    grad = data.grad.data
+                    device = grad.device
+                    if self._world_size > 0:
+                        idx_list = []
+                        grad_list = []
+                        for i in range(self._world_size):
+                            start = i * range_size
+                            end = (i + 1) * range_size \
+                                if (i + 1) * range_size < num_embeddings \
+                                else num_embeddings
+                            if i == 0:
+                                mask = idx < end
+                            elif i + 1 == self._world_size:
+                                mask = idx >= start
+                            else:
+                                mask = th.logical_and((idx >= start),(idx < end))
+                            idx_i = idx[mask]
+                            grad_i = grad[mask]
+
+                            if i == self._rank:
+                                update_embs[emb_name][0].append(idx_i)
+                                update_embs[emb_name][1].append(grad_i)
+                            else:
+                                emb.queues[i].put((emb_name,
+                                                   idx_i.to(th.device('cpu'), non_blocking=True).share_memory_(),
+                                                   grad_i.to(th.device('cpu'), non_blocking=True).share_memory_()))
```

Comment:
does this solution work for an asynchronous update?

Reply:
Not now. We can support that and it can be another PR.

Reply:
actually, my question is whether this solution can support async updates.

Reply:
You mean async training?
Current implementation can not support fully async updates. The information exchange can not be done async.